builtin/logical/database/rotation.go

```diff
@@ -0,0 +1,522 @@
+	}
+
+	// Store updated role information
+	lvr = time.Now()
+	input.Role.StaticAccount.LastVaultRotation = lvr
+	input.Role.StaticAccount.Password = password
+	output.RotationTime = lvr
+
+	entry, err := logical.StorageEntryJSON(databaseStaticRolePath+input.RoleName, input.Role)
+	if err != nil {
+		return output, err
+	}
+	if err := s.Put(ctx, entry); err != nil {
+		return output, err
+	}
+
+	// Cleanup WAL after successfully rotating and pushing new item on to queue
+	if err := framework.DeleteWAL(ctx, s, output.WALID); err != nil {
+		merr = multierror.Append(merr, err)
+		return output, merr
+	}
+
+	// The WAL has been deleted, return new setStaticAccountOutput without it
+	return &setStaticAccountOutput{RotationTime: lvr}, merr
+}
+
+// initQueue preforms the necessary checks and initializations needed to preform
+// automatic credential rotation for roles associated with static accounts. This
+// method verifies if a queue is needed (primary server or local mount), and if
+// so initializes the queue and launches a go-routine to periodically invoke a
+// method to preform the rotations.
+//
+// initQueue is invoked by the Factory method in a go-routine. The Factory does
+// not wait for success or failure of it's tasks before continuing. This is to
+// avoid blocking the mount process while loading and evaluating existing roles,
+// etc.
+func (b *databaseBackend) initQueue(ctx context.Context, conf *logical.BackendConfig) {
+	// Verify this mount is on the primary server, or is a local mount. If not, do
+	// not create a queue or launch a ticker. Both processing the WAL list and
+	// populating the queue are done sequentially and before launching a
+	// go-routine to run the periodic ticker.
+	replicationState := conf.System.ReplicationState()
+	if (conf.System.LocalMount() || !replicationState.HasState(consts.ReplicationPerformanceSecondary)) &&
+		!replicationState.HasState(consts.ReplicationDRSecondary) &&
+		!replicationState.HasState(consts.ReplicationPerformanceStandby) {
+		b.Logger().Info("initializing database rotation queue")
+
+		// Poll for a PutWAL call that does not return a "read-only storage" error.
+		// This ensures the startup phases of loading WAL entries from any possible
+		// failed rotations can complete without error when deleting from storage.
+	READONLY_LOOP:
+		for {
+			walID, err := framework.PutWAL(ctx, conf.StorageView, staticWALKey, &setCredentialsWAL{RoleName: "vault-readonlytest"})
+			if walID != "" {
+				defer framework.DeleteWAL(ctx, conf.StorageView, walID)
+			}
+			switch {
+			case err == nil:
+				break READONLY_LOOP
+			case err.Error() == logical.ErrSetupReadOnly.Error():
+				time.Sleep(10 * time.Millisecond)
+			default:
+				b.Logger().Error("deleting nil key resulted in error", "error", err)
+				return
+			}
+		}
+
+		// Load roles and populate queue with static accounts
+		b.populateQueue(ctx, conf.StorageView)
+
+		// Launch ticker
+		go b.runTicker(ctx, conf.StorageView)
```

Comment:
This seems to be in the context of the Factory call.  I'm not sure that is right but could be missing something on how this works.  I'm guessing you want to put this in the background that somehow gets triggered on shutdown.

Reply:
I'm not sure I follow, can you clarify here? "in the context of the Factory call" are you talking about the literal context `ctx` here? 

The `ctx` here has a `Done` channel, created in the Factory function:

- https://github.com/hashicorp/vault/blob/ea4f96a8240b6119d728ac18b84bbbe40c65dffd/builtin/logical/database/backend.go#L59

The done channel is closed in the `invalidateQueue` method, which is triggered by the Backends `clean` method:

- https://github.com/hashicorp/vault/blob/ea4f96a8240b6119d728ac18b84bbbe40c65dffd/builtin/logical/database/backend.go#L276

Reply:
Doesn't the factory context get cancelled once the factory method has completed?

Reply:
I traced it up to `vault/core.go`, I'm not seeing where this context gets cancelled, but I'm not 100% up to speed on core. In usage I have not seen the queue get shut down when I assume the Factory would have been completed. 

It's simple enough to create a new context and use that though, would you prefer that?

Reply:
Updated to create new context in 7e6ca8ac1